# Preprocessing and Exploritory Data Analysis

In [2]:
# Importing the necessary libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [7]:
# Load in the dataframe
df = pd.read_csv('../data/db_computers_2025_raw.csv')

/var/folders/_m/kfvhsj6515q3c7pdw8kq_qwc0000gn/T/ipykernel_98203/2794974568.py:2: DtypeWarning: Columns (6,7,8,9,13,14,17,18,19,20,21,22,24,25,26,27,28,29,30,32,36,37,40,41,42,46,49,50,51,53,59,60,62,63,67,70,73,75,76,77,78,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,98,100,103,104,105,126,127,128,129,130,132,133,134,135) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('../data/db_computers_2025_raw.csv')


## Data inpection & summary statistics 

### Initial examination

In [11]:
df.head()


Unnamed: 0                            Título             Precio_Rango  \
0           0     Apple MacBook Air 13" 2024 M3  1.026,53 € – 2.287,17 €   
1           1     Apple MacBook Air 15" 2024 M3  1.270,42 € – 2.672,25 €   
2           2   Apple MacBook Pro 14" M4 (2024)  1.731,44 € – 4.932,25 €   
3           3  ASUS TUF Gaming A15 FA506 (2024)    647,00 € – 1.031,71 €   
4           4         GigaByte Aero 16 OLED BKF  1.031,31 € – 2.069,99 €   

        Ofertas     Tipo de producto              Serie  \
0  200 ofertas:  Portátil multimedia  Apple MacBook Air   
1  230 ofertas:  Portátil multimedia  Apple MacBook Air   
2  136 ofertas:  Portátil multimedia  Apple MacBook Pro   
3   30 ofertas:      Portátil gaming    ASUS TUF Gaming   
4   29 ofertas:      Portátil gaming      Gigabyte Aero   

  Pantalla_Tamaño de la pantalla Pantalla_Diagonal de la pantalla  \
0                  13,6 pulgadas                        34,544 cm   
1                  15,3 pulgadas                        38,862 cm   
2                  14,2 pulgadas                        36,068 cm   
3                  15,6 pulgadas                        39,624 cm   
4                    16 pulgadas                         40,64 cm   

  Pantalla_Tecnología de la pantalla Pantalla_Luminosidad  ...  \
0                             Retina            500 cd/m²  ...   
1                             Retina            500 cd/m²  ...   
2                             Retina          1.000 cd/m²  ...   
3                            Full HD                  NaN  ...   
4                               UHD+            600 cd/m²  ...   

  Características especiales Almacenamiento_Formato audiovisual compatible  \
0                        NaN                                           NaN   
1                        NaN                                           NaN   
2                        NaN                                           NaN   
3                        NaN                                           NaN   
4                        NaN                                           NaN   

  Otras características_Modo de sonido  \
0                                  NaN   
1                                  NaN   
2                                  NaN   
3                                  NaN   
4                                  NaN   

  Propiedades de la carcasa_Superficie del material  \
0                                               NaN   
1                                               NaN   
2                                               NaN   
3                                               NaN   
4                                               NaN   

  Disco duro_Capacidad disco duro (GB)  \
0                                  NaN   
1                                  NaN   
2                                  NaN   
3                                  NaN   
4                                  NaN   

  Propiedades de la carcasa_Bahía externa de 3,5" Procesador_Placa base  \
0                                             NaN                   NaN   
1                                             NaN                   NaN   
2                                             NaN                   NaN   
3                                             NaN                   NaN   
4                                             NaN                   NaN   

  Otras características_Certificado RoHS Sistema operativo_Software incluido  \
0                                    NaN                                 NaN   
1                                    NaN                                 NaN   
2                                    NaN                                 NaN   
3                                    NaN                                 NaN   
4                                    NaN                                 NaN   

  Procesador_Procesadores compatibles  
0                                 NaN  
1                                 NaN  
2                    

Based on the head of the dataset, we can see that the dataset has large amount of collumns. This is most likley due to the different types of PC parts that have their own unique properties that do not apply to other categories.

In [12]:
df.tail()

Unnamed: 0                                           Título  \
8059        4027              Ankermann Work Home (4260561745772)   
8060        4028             Captiva Advanced Gaming I75\nI75-363   
8061        4029  Vibox VIII-49 Gaming PC i9-11900F / RTX 4070 Ti   
8062        4030  Sedatech UCCG810I2I1HF - Watercooling Gaming-PC   
8063        4031              Ankermann Gaming PC (4260561745598)   

     Precio_Rango    Ofertas Tipo de producto Serie  \
8059     249,00 €  1 oferta:    PC de oficina   NaN   
8060   1.129,99 €  1 oferta:        PC gaming   NaN   
8061   2.249,95 €  1 oferta:        PC gaming   NaN   
8062   3.269,90 €  1 oferta:        PC gaming   NaN   
8063   2.543,00 €  1 oferta:        PC gaming   NaN   

     Pantalla_Tamaño de la pantalla Pantalla_Diagonal de la pantalla  \
8059                            NaN                              NaN   
8060                            NaN                              NaN   
8061                            NaN                              NaN   
8062                            NaN                              NaN   
8063                            NaN                              NaN   

     Pantalla_Tecnología de la pantalla Pantalla_Luminosidad  ...  \
8059                                NaN                  NaN  ...   
8060                                NaN                  NaN  ...   
8061                                NaN                  NaN  ...   
8062                                NaN                  NaN  ...   
8063                                NaN                  NaN  ...   

     Características especiales Almacenamiento_Formato audiovisual compatible  \
8059                        NaN                                           NaN   
8060                        NaN                                           NaN   
8061                        NaN                                           NaN   
8062                        NaN                                           NaN   
8063                        NaN                                           NaN   

     Otras características_Modo de sonido  \
8059                                  NaN   
8060                                  NaN   
8061                                  NaN   
8062                                  NaN   
8063                                  NaN   

     Propiedades de la carcasa_Superficie del material  \
8059                                               NaN   
8060                                               NaN   
8061                                               NaN   
8062                                               NaN   
8063                                               NaN   

     Disco duro_Capacidad disco duro (GB)  \
8059                               500 GB   
8060                                  NaN   
8061                                  NaN   
8062                                  NaN   
8063                                  NaN   

     Propiedades de la carcasa_Bahía externa de 3,5" Procesador_Placa base  \
8059                                             NaN                   NaN   
8060                                             NaN                   NaN   
8061                                             NaN                   NaN   
8062                                             NaN                   NaN   
8063                                             NaN                   NaN   

     Otras características_Certificado RoHS  \
8059                                    NaN   
8060                                    NaN   
8061                                    NaN   
8062                                    NaN   
8063                                    NaN   

     Sistema operativo_Software incluido Procesador_Procesadores compatibles  
8059                                 NaN                                 NaN  
8060                                 NaN                                 NaN  
8061                                 NaN     

In [17]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8064 entries, 0 to 8063
Columns: 136 entries, Unnamed: 0 to Procesador_Procesadores compatibles
dtypes: float64(10), int64(1), object(125)
memory usage: 8.4+ MB


There are 8063 items in this dataset that we will have to deal with. An important part to look at is the data types. We have 10 Columns with type Float, 1 column with an integer type, and 125 object columns. These datatypes are important to help us understand what type of database is suitable for this dataset.

### Summary Statistics

In [19]:
df.describe()

Unnamed: 0  Sonido_Número de altavoces  \
count  8064.000000                 2938.000000   
mean   2015.500000                    2.477195   
std    1164.010282                    1.111884   
min       0.000000                    2.000000   
25%    1007.750000                    2.000000   
50%    2015.500000                    2.000000   
75%    3023.250000                    2.000000   
max    4031.000000                    8.000000   

       Otras características_Fecha de lanzamiento  \
count                                 3781.000000   
mean                                  2022.613065   
std                                      1.511238   
min                                   2012.000000   
25%                                   2022.000000   
50%                                   2023.000000   
75%                                   2024.000000   
max                                   2025.000000   

       Disco duro_Número de discos duros (instalados)  \
count                                     5598.000000   
mean                                         0.988567   
std                                          0.238087   
min                                          0.000000   
25%                                          1.000000   
50%                                          1.000000   
75%                                          1.000000   
max                                          3.000000   

       Alimentación_Número de celdas  Procesador_Número de hilos de ejecución  \
count                    2679.000000                              3102.000000   
mean                        3.530795                                16.374275   
std                         1.037607                                 7.292888   
min                         2.000000                                 1.000000   
25%                         3.000000                                12.000000   
50%                         3.000000                                16.000000   
75%                         4.000000                                20.000000   
max                         8.000000                                32.000000   

       RAM_Número de ranuras para memoria RAM  \
count                             1043.000000   
mean                                 2.478428   
std                                  1.081277   
min                                  1.000000   
25%                                  2.000000   
50%                                  2.000000   
75%                                  2.000000   
max                                  8.000000   

       Propiedades de la carcasa_Bahía interna de 2,5"  \
count                                       358.000000   
mean                                          1.304469   
std                                           0.603010   
min                                           1.000000   
25%                                           1.000000   
50%                                           1.000000   
75%                                           2.000000   
max                                           6.000000   

       Propiedades de la carcasa_Bahía interna de 3,5"  \
count                                       138.000000   
mean                                          1.557971   
std                                           0.704709   
min                                           1.000000   
25%                                           1.000000   
50%                                           1.000000   
75%                                           2.000000   
max                                           5.000000   

       Propiedades de la carcasa_Bahía externa de 5,25"  \
count                                         49.000000   
mean                                           1.632653   
std                                            0.528121   
min                                            1.000000   
25%                                            1.000

Since not every row has values for all the columns, the .describe function does not help us as much as it could. We may have to split the dataset into multiple datasets to get a better understanding of the data.

In [26]:
df.nunique()


Unnamed: 0                                         4032
Título                                             8053
Precio_Rango                                       6706
Ofertas                                             110
Tipo de producto                                     17
                                                   ... 
Propiedades de la carcasa_Bahía externa de 3,5"       2
Procesador_Placa base                                13
Otras características_Certificado RoHS                1
Sistema operativo_Software incluido                   5
Procesador_Procesadores compatibles                   4
Length: 136, dtype: int64

There seems to be almost zero non-unqiue features. There are 8064 entries and 8053 of them are unique. All have their own values for properties that differeniate eachother. This is helpful to know that there are almost zero duplicate PC products. Let check now for any duplicates.

In [29]:
df.duplicated().sum()

0

Despite the number of entries (8064) and unqiue titles (8053) only slightly differing, there are no exact duplicate rows. So there is no further duplicate data handling needed.

## Visualization

## Feature Cleaning

## Transformation

## Feature Engineering

## Validation & Consistency Checks